### PR TITLE
Refactor deployed_at and latest_commit_id code

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
I created two PR's https://github.com/forem/forem/pull/12149 and https://github.com/forem/forem/pull/12182 that worked on adding the Deployed at and the Latest commit ID to the admin interface. 

A [suggestion](https://github.com/forem/forem/pull/12149#discussion_r553075918) that came out of https://github.com/forem/forem/pull/12149 was to add a new model (not DB backed) like ForemInstance to which we add a bunch of class methods that encapsulate things that are used for Forem Instances and will change across the different instances. It made sense to me and so I've added that model. 

Aside: As per the comment, we could even add move dev_to? from SiteConfig to the new class since ForemInstance.dev_to? makes a lot more sense than SiteConfig.dev_to?. Same applied to the local? method for the same reason (ForemInstance.local?). This can be refactored and added at a later stage.

**TLDR:**

- This PR focuses on abstracting out the deployed_at and the lastest_commit_id into the ForemInstance model. 
- It also replaces the instances where we use `ENV['RELEASE_FOOTPRINT']` to use ForemInstace.deployed_at. 
- I've removed the set_release_key initializer in favor of the Forem Instance model which explicitly accesses the correct variable directly using ForemInstance Model 
- Finally, in the body we have some data keys that I rename to be more apt and inline with what these variables do

## Related Tickets & Documents
- https://github.com/forem/forem/pull/12149
- https://github.com/forem/forem/pull/12182

## QA Instructions, Screenshots, Recordings
It's a bit difficult to test these changes since they swap out an ENV variable for the ForemInstance model method. Some of these changes are also on the caching methods. 

However, you can play aroudn with it by changing the env values, restarting your app and making sure that the data-deployed-at and the data-latest-commit-id update.

### UI accessibility concerns?

None.

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?
Keep an eye out for any caching issues.

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/controllers/async_info_controller.rb [hunk 1]
```diff
@@ -29,7 +29,7 @@ def shell_version
     set_surrogate_key_header "shell-version-endpoint"
     # shell_version will change on every deploy.
     # *Technically* could be only on changes to assets and shell, but this is more fool-proof.
-    shell_version = ApplicationConfig["RELEASE_FOOTPRINT"] + SiteConfig.admin_action_taken_at.to_s
+    shell_version = ForemInstance.deployed_at + SiteConfig.admin_action_taken_at.to_s
```
Comment (Ridhwana) [inline app/controllers/async_info_controller.rb:32]: Using the ForemInstance.deployed_at method here as will set ApplicationConfig["RELEASE_FOOTPRINT"] if it exists.

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -174,7 +174,7 @@ def community_emoji
   end
 
   def release_adjusted_cache_key(path)
-    release_footprint = ApplicationConfig["RELEASE_FOOTPRINT"]
+    release_footprint = ForemInstance.deployed_at
```
Comment (Ridhwana) [inline app/helpers/application_helper.rb:177]: Using the newly created ForemInstance.deployed_at method here as will set ApplicationConfig["RELEASE_FOOTPRINT"] if it is present.

File: app/models/forem_instance.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class ForemInstance
```
Comment (Ridhwana) [inline app/models/forem_instance.rb:1]: A suggestion that came out of #12149 was to add a model (not DB backed) i.e. ForemInstance to which we add a bunch of class methods that encapsulate things that are used for Forem Instances and will change across the different instances. It made sense to start that model with a deployed_at and the latest_commit_id

File: app/models/forem_instance.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  def self.deployed_at
+    ApplicationConfig["RELEASE_FOOTPRINT"].presence || ENV["HEROKU_RELEASE_CREATED_AT"].presence
```
Comment (Ridhwana) [inline app/models/forem_instance.rb:3]: For the Forem.cloud environment RELEASE_FOOTPRINT will be set whereas for a Heroku environment HEROKU_RELEASE_CREATED_AT will be set

File: app/models/forem_instance.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+  end
+
+  def self.latest_commit_id
+    ApplicationConfig["FOREM_BUILD_SHA"].presence || ENV["HEROKU_SLUG_COMMIT"].presence
```
Comment (Ridhwana) [inline app/models/forem_instance.rb:7]: For the Forem.cloud environment FOREM_BUILD_SHA will be set whereas for a Heroku environment HEROKU_SLUG_COMMIT will be set

File: app/views/admin/admin_portals/index.html.erb [hunk 6]
```diff
@@ -2,11 +2,11 @@
   <div class="tag-card crayons-card p-4 m:p-6 m:pt-4 mb-4 flex flex-row relative">
     <div class="mr-8">
       <div class="crayons-field__description">Last deployed</div>
-      <div><%= ApplicationConfig["RELEASE_FOOTPRINT"].presence || ENV["HEROKU_RELEASE_CREATED_AT"].presence || "Not Available" %></div>
+      <div><%= ForemInstance.deployed_at || "Not Available" %></div>
     </div>
     <div>
       <div class="crayons-field__description">Latest Commit ID</div>
-      <div><%= ApplicationConfig["FOREM_BUILD_SHA"].presence || ENV["HEROKU_SLUG_COMMIT"].presence || "Not Available" %></div>
+      <div><%= ForemInstance.latest_commit_id || "Not Available" %></div>
```
Comment (Ridhwana) [inline app/views/admin/admin_portals/index.html.erb:9]: Change the display for the admin portal page to reuse the ForemInstance methos instance of explicitly referencing the variables

File: app/views/layouts/admin.html.erb [hunk 7]
```diff
@@ -26,7 +26,8 @@
 </head>
 <body class="admin default-header"
   data-honeybadger-key="<%= ApplicationConfig["HONEYBADGER_JS_API_KEY"] %>"
-  data-release-footprint="<%= ApplicationConfig["RELEASE_FOOTPRINT"] %>"
```
Comment (Ridhwana) [inline app/views/layouts/admin.html.erb:29]: **Confirmation Required** I'm not 100% sure if this key being referenced by an external app, so I need some confirmation here that it is fine to change the data key name. 

It made more sense to name it in accordance to what it references and consistent with the rest of the app. I've also added the latest commit id to the data keys on the body.

File: config/initializers/set_release_footprint.rb [hunk 8]
```diff
@@ -1,2 +0,0 @@
-# We will set RELEASE_FOOTPRINT in our Forem Cloud environment, or use HEROKU_SLUG_COMMIT if set (e.g. Heroku env)
-ENV["RELEASE_FOOTPRINT"] ||= ENV["HEROKU_RELEASE_CREATED_AT"]
```
Comment (Ridhwana) [inline config/initializers/set_release_footprint.rb:2]: **Confirmation required:** I removed this because we have explicitly set the variables in places where we use them. I could totally be missing an important part of the code, hence it would be helpful to get other developer eyes on this.

**Comment by Ridhwana** [comment]
This PR touches some of the caching stuff by using the new ForemInstance model. It's a pretty simple change and I've outlined some of my rationale for making some changes but would really appreciate a couple of reviews on this PR to ensure that I'm not missing a critical piece of information and possibly breaking some caching and other related stuff.

**Comment by Ridhwana** [comment]
Update: I'm looking into the failing test.

File: app/models/forem_instance.rb (hunk 4 above)
Comment (rhymes) [reply app/models/forem_instance.rb:3]: Since both are taken from static values that can never change during the lifetime of the app (you literally need to restart the app to change environment variables), I think the result of the `OR` can be frozen into a constant. We don't need to call a method each time:

`ForemInstance::DEPLOYED_AT` <- this value should never change during the lifetime of the deployed version

File: app/models/forem_instance.rb (hunk 5 above)
Comment (rhymes) [reply app/models/forem_instance.rb:7]: same thing about moving this to a frozen constant

**Review by rhymes — COMMENTED** [review]
Left notes about moving method calls to constants for constant values, I think `dev_to?` and `local?` should be methods instead

File: app/models/forem_instance.rb (hunk 4 above)
Comment (Ridhwana) [reply app/models/forem_instance.rb:3]: ✅

File: app/models/forem_instance.rb (hunk 5 above)
Comment (Ridhwana) [reply app/models/forem_instance.rb:7]: ✅

File: config/initializers/set_release_footprint.rb (hunk 8 above)
Comment (mstruve) [reply config/initializers/set_release_footprint.rb:2]: seems like dead code to me.

**Review by mstruve — APPROVED** [review]
This is fantastic!!!

File: app/models/forem_instance.rb (hunk 4 above)
Comment (citizen428) [reply app/models/forem_instance.rb:3]: Tbh, I still prefer the method, we could always cache it. Frequently reaching into constants defined by other models seems a bit off to me.

```ruby
def self.deployed_at
  @deployed_at ||=  ApplicationConfig["RELEASE_FOOTPRINT"].presence || ENV["HEROKU_RELEASE_CREATED_AT"].presence
end
```

Comment (Ridhwana) [reply app/models/forem_instance.rb:3]: I spent time considering the pro's and cons of each approach, and I don't think theres a right or wrong answer here (especially using the cached method strategy), but it's rather a matter of preference. 

I had put this PR on hold yesterday evening after struggling with testing the variables. After discussion with Rhymes, we had concluded that it was testing essentially an "or" statement and so it would be relatively safe to remove the tests if we weren't able to stub it out.

With a fresh mind this morning I tackled it again and after reading the comment above I've decided to move it back to a method with caching. This will help us to keep these tests in place as well.

**Comment by Ridhwana** [comment]
Moved this PR from Draft to Ready for Review. Addressed all feedback and fixed the tests. One thing to watch out for when merging and deploying this PR is any cache issues.

**Review by rhymes — APPROVED** [review]
LGTM
